hydradx/tests/test_omnipool_agents.py

```diff
@@ -168,3 +168,38 @@ def test_omnipool_LP(omnipool: oamm.OmnipoolState):
     final_state = events[-1]
     if final_state.agents['agent'].holdings['HDX'] != 0 or events[-2].agents['agent'].holdings['HDX'] == 0:
         raise AssertionError('HDX not invested at the right time.')
+
+
+def test_agent_copy():
+    import random
+    init_agent = Agent(holdings={'HDX': 100, 'USD': 100}, share_prices={'HDX': 1, 'USD': 1})
+    def randomize_object(obj):
+        if type(obj) == dict:
+            return {key: randomize_object(obj[key]) for key in obj}
+        elif type(obj) == list:
+            return [randomize_object(item) for item in obj]
+        elif type(obj) in [int, float]:
+            return random.random() * obj
+        elif type(obj) == str:
+            return obj + str(random.random())
+        elif obj is None:
+            return random.random()
+        else:
+            for prop in obj.__dict__:
+                setattr(obj, prop, randomize_object(getattr(obj, prop)))
+
+    mod_agent = init_agent.copy()
+    randomize_object(mod_agent)
+
+    copy_agent = mod_agent.copy()
+    copy_init_agent = init_agent.copy()
```

Comment:
I think you should be able to just randomize the initial agent, copy that, and then compare?

Reply:
Randomizing is necessary, because otherwise we could miss an attribute if it defaults to 0, for example. It's a way of ensuring that 1. we're using all the attributes 2. they will not match unless copied correctly and 3. 1 & 2 are automated.

I think you're right about not needing all those objects, though. Now that the random one is there I can delete the original.